File: couch-server/src/main/java/com/fronde/server/services/transmitter/TransmitterBaseService.java
package com.fronde.server.services.transmitter;

import com.fronde.server.domain.TransmitterEntity;
import com.fronde.server.domain.response.DeleteByIdCheckDTO;
import com.fronde.server.domain.response.PagedResponse;
import com.fronde.server.domain.response.Response;
import com.fronde.server.services.common.ObjectIdFactory;
import com.fronde.server.services.revision.RevisionService;
import com.fronde.server.services.validation.ValidationMessage;
import com.fronde.server.services.validation.ValidationService;
import com.fronde.server.utils.BaseEntityUtils;
import com.fronde.server.utils.CSVExportUtils;
import java.util.Arrays;
import java.util.List;
import javax.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;


@Component
public abstract class TransmitterBaseService {

  @Autowired
  protected TransmitterRepository repository;

  @Autowired
  protected ValidationService validationService;

  @Autowired
  protected RevisionService revisionService;

  @Autowired
  protected CSVExportUtils exportUtils;

  @Autowired
  protected ObjectIdFactory objectIdFactory;

  public TransmitterEntity findById(String docId) {
    // using orElse instead of get so we return null if the entity is not found
    return repository.findById(docId).orElse(null);
  }

  public PagedResponse<TransmitterEntity> search(TransmitterCriteria criteria) {
    return repository.findByCriteria(criteria);
  }

  public Response<TransmitterEntity> save(TransmitterEntity entity) {
    List<ValidationMessage> messages = validationService.validateEntity(entity);
    if (messages.size() == 0) {
      if (entity.getId() == null) {
        entity.setId(objectIdFactory.create());
      }

      if (entity.getDocType() == null) {
        entity.setDocType("Transmitter");
      }

      BaseEntityUtils.setModifiedFields(entity);

      revisionService.createRevision(entity);
      entity = repository.save(entity);
    }

    Response<TransmitterEntity> response = new Response<>(entity);
    response.setMessages(messages);
    return response;
  }

  public void deleteById(String id) {
    DeleteByIdCheckDTO deleteByIdCheckDTO = deleteByIdCheck(id);
    if (deleteByIdCheckDTO != null && deleteByIdCheckDTO.getDeleteOk()) {
      repository.deleteById(id);
    } else {
      throw new RuntimeException("Delete check failed for: " + id);
    }
  }

  public abstract DeleteByIdCheckDTO deleteByIdCheck(String docId);

  public void export(TransmitterCriteria criteria, HttpServletResponse response) {
    List<String> header = Arrays.asList(
        "Tx Id", "Status", "Island", "Channel", "Mort Type", "Bird", "Life Expectancy",
        "Rigging");
    List<String> props = Arrays.asList(
        "txId", "status", "island", "channel", "mortType", "birdId", "lifeExpectancy",
        "rigging");
    criteria.setPageSize(exportUtils.getMaxRows());
    criteria.setPageNumber(1);
    PagedResponse<TransmitterEntity> pr = this.search(criteria);
    exportUtils.export(response, pr, header, props, "transmitter");
  }

}

File: angular-client/src/app/transmitter/transmitter-base.service.ts
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { TransmitterEntity } from '../domain/transmitter.entity';
import { PagedResponse } from '../domain/response/paged-response';
import { DeleteByIdCheckDto } from '../domain/response/delete-by-id-check-dto';
import { Response } from '../domain/response/response';

export abstract class TransmitterBaseService {
  constructor(protected http: HttpClient) {}

  public search(criteria: any): Observable<PagedResponse> {
    return this.http.post<PagedResponse>('/api/transmitter/search', criteria);
  }

  public findById(docId: string): Observable<TransmitterEntity> {
    return this.http.get<TransmitterEntity>('/api/transmitter/' + docId).pipe(
      map((response: TransmitterEntity) => {
        response.modifiedTime =
          response.modifiedTime === null
            ? null
            : new Date(response.modifiedTime);
        return response;
      })
    );
  }

  public save(
    entity: TransmitterEntity
  ): Observable<Response<TransmitterEntity>> {
    return this.http.post<Response<TransmitterEntity>>(
      '/api/transmitter/',
      entity
    );
  }

  public delete(docId: string): Observable<any> {
    return this.http.delete<any>('/api/transmitter/' + docId);
  }

  public deleteByIdCheck(docId: string): Observable<DeleteByIdCheckDto> {
    return this.http.delete<any>('/api/transmitter/' + docId + '/check');
  }
}
